dbr quant function fusion [1/x]: record matches for functions

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71781 dbr quant function fusion [2/x]: use fusion for observation and inference
* #71780 fx quant: add fusion matching for operator.add and torch.relu
* **#71764 dbr quant function fusion [1/x]: record matches for functions**
* #71551 dbr quant: record dag of non-quantizeable ops
* #71324 dbr quant: refactor first_call hooks to be standalone
* #71312 dbr quant: rename seen_op_info to seen_q_op_info

Summary:

For DBR quant, adds the code for matching seen ops to function fusion
patterns. After we have the full DAG, we have a separate pass over the
dag and add matched fusion patterns to the seen op data structure.

This is the first PR in the stack which implements matching and
recording the match results. Future PRs in this stack will use
the match results to modify observer insertion and inference.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_fusion_functions
```

Differential Revision: [D33775098](https://our.internmc.facebook.com/intern/diff/D33775098)

vkuzo committed:
dbr quant function fusion [1/x]: record matches for functions

Summary:

For DBR quant, adds the code for matching seen ops to function fusion
patterns. After we have the full DAG, we have a separate pass over the
dag and add matched fusion patterns to the seen op data structure.

This is the first PR in the stack which implements matching and
recording the match results. Future PRs in this stack will use
the match results to modify observer insertion and inference.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_fusion_functions
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33775098).

vkuzo committed:
Update on "dbr quant function fusion [1/x]: record matches for functions"


Summary:

For DBR quant, adds the code for matching seen ops to function fusion
patterns. After we have the full DAG, we have a separate pass over the
dag and add matched fusion patterns to the seen op data structure.

This is the first PR in the stack which implements matching and
recording the match results. Future PRs in this stack will use
the match results to modify observer insertion and inference.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_fusion_functions
```

Differential Revision: [D33775098](https://our.internmc.facebook.com/intern/diff/D33775098)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33775098).

jerryzh168 commented on torch/ao/quantization/_dbr/fusion.py:
```diff
@@ -32,25 +32,21 @@ def get_module_fusion_fqns(
                 if is_match:
+                    cur_fqns = [seen_q_op_info.fqn]
+                    cur_seen_q_op_info = seen_q_op_info
+                    for element in fusion_pattern[:-1]:
```
nit: looks like element is not used

vkuzo committed:
Update on "dbr quant function fusion [1/x]: record matches for functions"


Summary:

For DBR quant, adds the code for matching seen ops to function fusion
patterns. After we have the full DAG, we have a separate pass over the
dag and add matched fusion patterns to the seen op data structure.

This is the first PR in the stack which implements matching and
recording the match results. Future PRs in this stack will use
the match results to modify observer insertion and inference.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_fusion_functions
```

Differential Revision: [D33775098](https://our.internmc.facebook.com/intern/diff/D33775098)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33775098).
